Bug fix: Welcome Message not getting triggered in Dev and Kek Vonal Staging

acedeywin opened the pull request:
<!-- Tag the primary responsible for reviewing this PR. If in doubt who can take this, ask first. -->
Primary reviewer:

## Description
<!--
- What this pull request does.
- Bug fix, new feature, documentation change, etc.
-->
- Bug fix: Welcome Message not getting triggered in Dev and Kek Vonal Staging

### Checklist
- [x] Corresponding issue has been opened
- [N/A] New tests added
- [N/A] Feature flags added
- [N/A] Strings are localized
- [N/A] Tested for chat contacts
- [N/A] Tested for call contacts

### Related Issues
Fixes CHI-1768

### Verification steps
<!--
Describe how to validate your changes.
- Include screen shots if applicable.
- Note if migrations are required.
-->

stephenhand reviewed (commented):
Can you explain what you changed in this PR - the diffs are not very clean and it's very difficult to see what you changed inside the try catch blocks

acedeywin commented:
> Can you explain what you changed in this PR - the diffs are not very clean, and it's very difficult to see what you changed inside the try-catch blocks

The welcome message not getting triggered because the `getUnknownOption` function was breaking, and the error was not correctly handled. So the `try-catch` block was added to handle the error graciously and not break the entire code.

The main error was that `getUnknownOption` couldn't determine a valid unknown option for key. In this case, `unknownOption` didn't have the right value or `e.value === 'Unknown'` does not exist within the `json` files for Kek Vonal (Line 31 in prepopulateForm.ts). So I did a bit of a refactor on Line 31 in prepopulateForm.ts and added values for `unknownOption` in some of the `json` files.

stephenhand commented on plugin-hrm-form/src/utils/setUpActions.ts:
```diff
@@ -206,7 +206,7 @@ export const afterAcceptTask = (featureFlags: FeatureFlags, setupObject: SetupOb
 
   // If this is the first counsellor that gets the task, say hi
-  if (TaskHelper.isChatBasedTask(task) && !TransferHelpers.hasTransferStarted(task)) {
+  if (TaskHelper.isChatBasedTask(task)) {
```
What is this change for? Is it part of you PR or a change from elsewhere?

acedeywin replied:
@stephenhand, Ooh, I removed it while troubleshooting. Thank you for spotting it. I will add it back.

stephenhand replied:
You might have added back to the wrong if statement?

acedeywin replied:
Fixed.

stephenhand reviewed (changes requested):
Please revert all the `try .. catch ... log` blocks. Swallowing errors is poor practice and should only be used in exceptional circumstances.

We should have proper error reporting for when prepopulation fails, i.e. catch the error at the top level and display an error to the user that there was a problem prepopulating the form, but that is out of scope for this ticket - for now, letting the errors throw & stopping the welcome message is better than swallowing them & hiding issues

acedeywin commented:
> Please revert all the `try .. catch ... log` blocks. Swallowing errors is poor practice and should only be used in exceptional circumstances.
> 
> We should have proper error reporting for when prepopulation fails, i.e. catch the error at the top level and display an error to the user that there was a problem prepopulating the form, but that is out of scope for this ticket - for now, letting the errors throw & stopping the welcome message is better than swallowing them & hiding issues

@stephenhand, I have fixed another bug in the `getAnswerOrUnknown` function. In the Kek Vonal case, the `key` (gender and age) is undefined since users don't answer any preEngagement question that involves it. So since `key` is undefined, the `isUnknownAnswer` variable throws an error because `isUnknownAnswer` only checks for two conditions - `answers[key].error` or `answers[key].answer === unknown`. So I added a third condition, `answers[key] === undefined`, which fixed the issue.

stephenhand commented on plugin-hrm-form/src/utils/prepopulateForm.ts:
```diff
@@ -72,7 +74,7 @@ const getAnswerOrUnknown = (
   if (itemDefinition.type === 'select') {
     const unknown = getUnknownOption(key, definition);
-    const isUnknownAnswer = answers[key].error || answers[key].answer === unknown;
+    const isUnknownAnswer = answers[key] === undefined || answers[key].error || answers[key].answer === unknown;
```
```suggestion
    const isUnknownAnswer = !answers[key] || answers[key].error || answers[key].answer === unknown;
```

acedeywin replied:
Done!

stephenhand commented on plugin-hrm-form/src/utils/setUpActions.ts:
```diff
@@ -198,15 +198,15 @@ export const afterAcceptTask = (featureFlags: FeatureFlags, setupObject: SetupOb
   const { task } = payload;
 
-  if (TaskHelper.isChatBasedTask(task)) {
+  if (TaskHelper.isChatBasedTask(task) && !TransferHelpers.hasTransferStarted(task)) {
```
See below

stephenhand commented on plugin-hrm-form/src/utils/mappers.ts:
```diff
@@ -72,23 +72,28 @@ export const mapChannelForInsights = (channel: string) => {
 
 export const mapAge = (ageOptions: string[]) => (age: string) => {
-  const ageInt = parseInt(age, 10);
+  try {
```
Can we remove this try ... catch ... log as well? Or is there a specific reason you are keeping it?

acedeywin replied:
Done!

stephenhand reviewed (approved):
LGTM
